rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,926 @@
+        the last dimension sizes across all input tensors, with sparse tensors instead
+        contributing `config[DENSE_DIMENSION][attribute]` units each.
+
+    Raises:
+        A `TFLayerConfigException` if no feature signatures are provided.
+
+    Attributes:
+        output_units: The last dimension size of the layer's output.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        feature_type: Text,
+        feature_type_signature: List[FeatureSignature],
+        config: Dict[Text, Any],
+    ) -> None:
+        if not feature_type_signature:
+            raise TFLayerConfigException(
+                "The feature type signature must contain some feature signatures."
+            )
+
+        super().__init__(
+            name=f"concatenate_sparse_dense_features_{attribute}_{feature_type}"
+        )
+
+        self._check_sparse_input_units(feature_type_signature)
+
+        self.output_units = self._calculate_output_units(
+            attribute, feature_type_signature, config
+        )
+
+        # Prepare dropout and sparse-to-dense layers if any sparse tensors are expected
+        self._tf_layers = {}
+        if any([signature.is_sparse for signature in feature_type_signature]):
+            self._prepare_layers_for_sparse_tensors(attribute, feature_type, config)
+
+    def _check_sparse_input_units(
+        self, feature_type_signature: List[FeatureSignature]
+    ) -> None:
+        """Checks that all sparse features have the same last dimension size."""
+        sparse_units = [
+            feature_sig.units
+            for feature_sig in feature_type_signature
+            if feature_sig.is_sparse
+        ]
+        if len(set(sparse_units)) > 1:
+            raise TFLayerConfigException(
+                f"All sparse features must have the same last dimension size but found "
+                f"different sizes: {set(sparse_units)}."
+            )
+
+    def _prepare_layers_for_sparse_tensors(
+        self, attribute: Text, feature_type: Text, config: Dict[Text, Any],
+    ) -> None:
+        """Sets up sparse tensor pre-processing before combining with dense ones."""
+        # For optionally applying dropout to sparse tensors
+        if config[SPARSE_INPUT_DROPOUT]:
+            self._tf_layers["sparse_dropout"] = layers.SparseDropout(
+                rate=config[DROP_RATE]
+            )
+
+        # For converting sparse tensors to dense
+        self._tf_layers["sparse_to_dense"] = layers.DenseForSparse(
+            name=f"sparse_to_dense.{attribute}_{feature_type}",
+            units=config[DENSE_DIMENSION][attribute],
+            reg_lambda=config[REGULARIZATION_CONSTANT],
+        )
+
+        # For optionally apply dropout to sparse tensors after they're converted to
+        # dense tensors.
+        if config[DENSE_INPUT_DROPOUT]:
+            self._tf_layers["dense_dropout"] = tf.keras.layers.Dropout(
```

Comment:
I think if these strings that are used multiple times throughout classes, such as `dense_dropout`, `sparse_dropout`, `sparse_to_dense`, etc.,  would receive some constants then code would be less error prone in case of renaming in the future

Reply:
I agree that using constants in general can be less error-prone, but I'm also unsure if it's needed here. Each of these strings is used:
- only within one class
- in only 2 places
Additionally, I think this extremely unlikely to change -- I struggle to imagine the changes that would make one of these strings used in more places. @twerkmeister does this change your opinion? 😅  Also, if I went ahead and switched to using constants, do you think I should define them within their respective classes, or globally for the whole module?

Reply:
I think within their respective classes would be the right place. I think it would be nice, but not a must have if you think it doesn't add value.

Reply:
I decided to follow your advice and turn the strings into constants. Even if in this case is doesn't add much value, it sets an example in the research codebase.